### PR TITLE
001 Job Roles

### PR DESCRIPTION
**Pull request by Adnanhoq**
Job_Roles Tables (Job_Locations, Job_Bands, Job_Roles, Job_Location_Connector)

### CODE REVIEW

File: src/main/resources/db/migrations/V2_Populate_Jobs_Data.sql [hunk 1]
```diff
@@ -0,0 +1,1182 @@
...
+
+assisting in their career development.",
+
+"https://kainossoftwareltd.sharepoint.com/people/Job%20Specifications/Forms/AllItems.aspx?id=%2Fpeople%2FJob%20Specifications%2FBusiness%20Development%20and%20Marketing%2FInside%20Sales%2FJob%20profile%20%2D%20Sales%20Development%20Consultant%20%28C%29%2Epdf&parent=%2Fpeople%2FJob%20Specifications%2FBusiness%20Development%20and%20Marketing%2FInside%20Sales&p=true&ga=1",
+
+"Business Development and Marketing",
+
+3,
+
+TIMESTAMP("2024-11-12",  "11:00:00"),
```
Comment (Leebee1392) [inline src/main/resources/db/migrations/V2_Populate_Jobs_Data.sql]: What's going on with your formatting?

**Review by Leebee1392 — CHANGES_REQUESTED** [review]
Migrations don't work when running flyway locally

File: src/main/java/org/example/controllers/JobRoleController.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    public JobRoleController(final JobRoleService jobRoleService) {
+        this.jobRoleService = jobRoleService;
+    }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+//    @RolesAllowed({UserRole.ADMIN, UserRole.USER})
+    @ApiOperation(
+            value = "Returns all Job Roles",
+            authorizations = @Authorization(value = HttpHeaders.AUTHORIZATION),
```
Comment (Leebee1392) [inline src/main/java/org/example/controllers/JobRoleController.java]: Remove authorisation from here as it's not part of this ticket

File: src/main/java/org/example/controllers/JobRoleController.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    JobRoleService jobRoleService;
+
+    public JobRoleController(final JobRoleService jobRoleService) {
+        this.jobRoleService = jobRoleService;
+    }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+//    @RolesAllowed({UserRole.ADMIN, UserRole.USER})
```
Comment (Leebee1392) [inline src/main/java/org/example/controllers/JobRoleController.java]: Remove comment

File: src/main/java/org/example/controllers/JobRoleController.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    @ApiOperation(
+            value = "Returns all Job Roles",
+            authorizations = @Authorization(value = HttpHeaders.AUTHORIZATION),
+            response = JobRole.class
+    )
+    public Response getJobRole() throws SQLException {
```
Comment (Leebee1392) [inline src/main/java/org/example/controllers/JobRoleController.java]: Should be plural

File: src/main/java/org/example/daos/JobRoleDao.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JobRoleDao {
+
+    public List<JobRole> getAllJobRolesFull() throws SQLException {
```
Comment (Leebee1392) [inline src/main/java/org/example/daos/JobRoleDao.java]: Please rename this, think of the penguins.

File: src/main/java/org/example/exceptions/Entity.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package org.example.exceptions;
+
+public enum Entity {
+
+    JOBROLERESPONSE("JobRoleResponse");
```
Comment (Leebee1392) [inline src/main/java/org/example/exceptions/Entity.java]: Is there a reason this is here?

File: src/main/java/org/example/models/JobRole.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
+package org.example.models;
+
+import java.util.Date;
+
+public class JobRole {
+    private int roleId;
+    private String roleName;
+    private String description;
+    private String responsibilities;
+    public String locations;
```
Comment (Leebee1392) [inline src/main/java/org/example/models/JobRole.java]: Shouldn't this be an enum as it's an enum in the database

File: src/main/java/org/example/models/JobRole.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    private String linkToJobSpec;
+    private String capability;
```
Comment (Leebee1392) [inline src/main/java/org/example/models/JobRole.java]: Shouldn't this be an enum as it's an enum in the database

File: src/main/java/org/example/models/JobRole.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    private String band;
```
Comment (Leebee1392) [inline src/main/java/org/example/models/JobRole.java]: Shouldn't this be an enum as it's an enum in the database

File: src/main/java/org/example/models/JobRole.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    private Date closingDate;
+    private boolean status;
+    private int positionsAvailable;
+
+    // Removed capability, status, positionsAvailable
+    // from constructor due to 7 parameters limit.
+    public JobRole(
```
Comment (Leebee1392) [inline src/main/java/org/example/models/JobRole.java:22]: As you haven't made anything required in the database table, you can't expect the constructor to have all the values

File: src/main/java/org/example/models/JobRoleResponse.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+public class JobRoleResponse {
+    private int roleId;
+    private String roleName;
+    private String locations;
+    private String capability;
+    private String band;
+    private Date closingDate;
+
+    public JobRoleResponse(final int roleId, final String roleName,
+                   final String locations,
```
Comment (Leebee1392) [inline src/main/java/org/example/models/JobRoleResponse.java]: Should not be plural as it's a string (should be an enum)

File: src/main/java/org/example/models/JobRoleResponse.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                   final String capability,
+                   final String band, final Date closingDate) {
```
Comment (Leebee1392) [inline src/main/java/org/example/models/JobRoleResponse.java]: Shouldn't this be an enum as it's an enum in the database

Comment (Leebee1392) [inline src/main/java/org/example/models/JobRoleResponse.java]: Formatting is off

File: src/main/java/org/example/validators/JobRoleValidator.java [hunk 13]
```diff
@@ -0,0 +1,4 @@
+package org.example.validators;
+
+public class JobRoleValidator {
```
Comment (Leebee1392) [inline src/main/java/org/example/validators/JobRoleValidator.java]: Where's the code?

File: src/test/java/org/example/controller/JobRoleControllerTest.java [hunk 14]
```diff
@@ -0,0 +1,6 @@
+package org.example.controller;
+
+public class JobRoleControllerTest {
```
Comment (Leebee1392) [inline src/test/java/org/example/controller/JobRoleControllerTest.java]: Where is the code?

File: src/test/java/org/example/service/JobRoleServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+import org.mockito.Mockito;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@ExtendWith(MockitoExtension.class)
+public class JobRoleServiceTest {
```
Comment (Leebee1392) [inline src/test/java/org/example/service/JobRoleServiceTest.java:29]: Missing some test cases

File: src/test/java/org/example/validator/JobRoleValidatorTest.java [hunk 16]
```diff
@@ -0,0 +1,4 @@
+package org.example.validator;
+
+public class JobRoleValidatorTest {
```
Comment (Leebee1392) [inline src/test/java/org/example/validator/JobRoleValidatorTest.java]: I am confused - where is the code?

**Review by Leebee1392 — CHANGES_REQUESTED** [review]
I don't think this meets the first acceptance criteria. Please have a look again.

File: src/main/java/org/example/exceptions/Entity.java (hunk 6 above)
Comment (Adnanhoq) [reply src/main/java/org/example/exceptions/Entity.java]: Used in the JobRoleService for if JobRoleResponses is Empty

File: src/main/java/org/example/validators/JobRoleValidator.java (hunk 13 above)
Comment (Adnanhoq) [reply src/main/java/org/example/validators/JobRoleValidator.java]: I made the file for the testers

File: src/test/java/org/example/controller/JobRoleControllerTest.java (hunk 14 above)
Comment (Adnanhoq) [reply src/test/java/org/example/controller/JobRoleControllerTest.java]: Made the file for the testers

File: src/test/java/org/example/validator/JobRoleValidatorTest.java (hunk 16 above)
Comment (Adnanhoq) [reply src/test/java/org/example/validator/JobRoleValidatorTest.java]: Made the file for the testers

File: src/main/java/org/example/controllers/JobRoleController.java (hunk 3 above)
Comment (Adnanhoq) [reply src/main/java/org/example/controllers/JobRoleController.java]: Fixed

File: src/main/java/org/example/controllers/JobRoleController.java (hunk 4 above)
Comment (Adnanhoq) [reply src/main/java/org/example/controllers/JobRoleController.java]: Fixed

File: src/main/java/org/example/daos/JobRoleDao.java (hunk 5 above)
Comment (Adnanhoq) [reply src/main/java/org/example/daos/JobRoleDao.java]: made it getAllJobRoles

File: src/main/java/org/example/controllers/JobRoleController.java (hunk 2 above)
Comment (Adnanhoq) [reply src/main/java/org/example/controllers/JobRoleController.java]: Fixed

File: src/main/java/org/example/models/JobRoleResponse.java (hunk 12 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRoleResponse.java]: Fixed

File: src/test/java/org/example/service/JobRoleServiceTest.java (hunk 15 above)
Comment (Adnanhoq) [reply src/test/java/org/example/service/JobRoleServiceTest.java:29]: Added

File: src/main/java/org/example/models/JobRole.java (hunk 7 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRole.java]: No its returned as a string

File: src/main/java/org/example/models/JobRole.java (hunk 8 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRole.java]: No its returned as a string

File: src/main/java/org/example/models/JobRole.java (hunk 9 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRole.java]: No its returned as a string

File: src/main/java/org/example/models/JobRoleResponse.java (hunk 11 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRoleResponse.java]: its a string of enums

File: src/main/java/org/example/models/JobRoleResponse.java (hunk 12 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRoleResponse.java]: No its returned as a string

File: src/main/java/org/example/daos/JobRoleDao.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.List;
+
+public class JobRoleDao {
+
+    public List<JobRole> getAllJobRoles() throws SQLException {
+        List<JobRole> jobRoles = new ArrayList<>();
+        try (Connection connection = DatabaseConnector.getConnection()) {
+            Statement statement = connection.createStatement();
+
+            ResultSet resultSet = statement.executeQuery(
```
Comment (Leebee1392) [inline src/main/java/org/example/daos/JobRoleDao.java:26]: This query needs to be updated

File: src/main/java/org/example/models/JobRole.java (hunk 7 above)
Comment (Leebee1392) [reply src/main/java/org/example/models/JobRole.java]: Your database design has enums so it should match your code no matter what the database returns

File: src/main/java/org/example/models/JobRole.java (hunk 8 above)
Comment (Leebee1392) [reply src/main/java/org/example/models/JobRole.java]: Your database design has enums so it should match your code no matter what the database returns

File: src/main/java/org/example/models/JobRole.java (hunk 9 above)
Comment (Leebee1392) [reply src/main/java/org/example/models/JobRole.java]: Your database design has enums so it should match your code no matter what the database returns

File: src/main/java/org/example/models/JobRole.java (hunk 10 above)
Comment (Leebee1392) [reply src/main/java/org/example/models/JobRole.java:22]: This hasn't been completed. Don't ignore comments please :)

File: src/main/java/org/example/models/JobRoleResponse.java [hunk 18]
```diff
@@ -0,0 +1,74 @@
+package org.example.models;
+
+import java.util.Date;
+
+public class JobRoleResponse {
```
Comment (Leebee1392) [inline src/main/java/org/example/models/JobRoleResponse.java:6]: Why do we have this?

File: src/main/java/org/example/validators/JobRoleValidator.java (hunk 13 above)
Comment (Leebee1392) [reply src/main/java/org/example/validators/JobRoleValidator.java]: No, they can make their own files

File: src/test/java/org/example/controller/JobRoleControllerTest.java (hunk 14 above)
Comment (Leebee1392) [reply src/test/java/org/example/controller/JobRoleControllerTest.java]: No, they can make their own files.

File: src/test/java/org/example/validator/JobRoleValidatorTest.java (hunk 16 above)
Comment (Leebee1392) [reply src/test/java/org/example/validator/JobRoleValidatorTest.java]: No, they can make their own files.

**Review by Leebee1392 — CHANGES_REQUESTED** [review]
Are you sure this meets requirements of the ticket?

File: src/main/java/org/example/daos/JobRoleDao.java [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+
+public class JobRoleDao {
+
+    public List<JobRole> getAllJobRoles() throws SQLException {
+        List<JobRole> jobRoles = new ArrayList<>();
+        try (Connection connection = DatabaseConnector.getConnection()) {
+//            assert connection != null;
```
Comment (Leebee1392) [inline src/main/java/org/example/daos/JobRoleDao.java]: Remove comment

File: src/main/java/org/example/enums/JobBands.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
+package org.example.enums;
+
+public enum JobBands {
```
Comment (Leebee1392) [inline src/main/java/org/example/enums/JobBands.java]: It needs to be JobBand

File: src/main/java/org/example/enums/Locations.java [hunk 21]
```diff
@@ -0,0 +1,57 @@
+package org.example.enums;
+
+public enum Locations {
```
Comment (Leebee1392) [inline src/main/java/org/example/enums/Locations.java]: Should be Location

File: src/main/java/org/example/models/JobRoleResponse.java (hunk 18 above)
Comment (Leebee1392) [reply src/main/java/org/example/models/JobRoleResponse.java:6]: Can you please answer my question and not just resolve the comment

File: src/main/java/org/example/models/JobRole.java (hunk 10 above)
Comment (Leebee1392) [reply src/main/java/org/example/models/JobRole.java:22]: This still hasn't been done

**Review by Leebee1392 — CHANGES_REQUESTED** [review]

File: src/main/java/org/example/enums/Locations.java (hunk 21 above)
Comment (Adnanhoq) [reply src/main/java/org/example/enums/Locations.java]: Fixed

File: src/main/java/org/example/enums/JobBands.java (hunk 20 above)
Comment (Adnanhoq) [reply src/main/java/org/example/enums/JobBands.java]: Fixed

File: src/main/java/org/example/daos/JobRoleDao.java (hunk 19 above)
Comment (Adnanhoq) [reply src/main/java/org/example/daos/JobRoleDao.java]: Done

File: src/main/java/org/example/models/JobRoleResponse.java (hunk 18 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRoleResponse.java:6]: i don't understand. If you mean the imports, we are importing List as there are numerous Location enums
and Date is for the closingdate

File: src/main/java/org/example/models/JobRole.java (hunk 10 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRole.java:22]: Fixed it by making it NOT NULL in database

File: src/main/java/org/example/models/JobRoleResponse.java (hunk 18 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRoleResponse.java:6]: FormattingLocations allows us to convert from List<String> into a String

File: README.md
Comment (jemorak) [inline README.md:1]: Caution: potential read me conflicts - because of my stupid mistake

Comment (Adnanhoq) [reply README.md:1]: No worries, I'll fix it when you merge

File: src/main/java/org/example/models/JobRole.java (hunk 9 above)
Comment (Adnanhoq) [reply src/main/java/org/example/models/JobRole.java]: Fixed

**Review by jemorak — APPROVED** [review]

**Review by Leebee1392 — APPROVED** [review]

**Review by Leebee1392 — APPROVED** [review]

**Review by Leebee1392 — APPROVED** [review]

**Pull request merged**